### ISSUE TITLE
False positive "UnnecessaryComparison" with string split and unpack operator

### ISSUE THREAD

**Issue by EnderASz** [microsoft/pyright#7611]
I have some code which uses str.split() function and I want to get some values from that string. If split provides not enough values, I want use just `None` value instead. Below I have a code snippet of it. Problem appears when I'm trying to do something with these variables. Here I expect that:
- `a` is always a string (`split` always returns at least one empty string)
- `b` is a string or None

Pyright reports that `b is None` check is unnecessary, because it thinks `b` is always a string - it's not a true.

Probably problem is in type checking of unpack operation - it doesn't properly interprets what types can appear here.

```python
a, b, *_ = *"nanan".split(), None

if b is None:  # error: Condition will always evaluate to False since the types "LiteralString" and "None" have no overlap (reportUnnecessaryComparison)
    ...
```

If this is intentional behaviour - why? And what is a better solution for such a case.

**VS Code extension or command-line**
Both reports the same - `Pylance` as VS Code extension and CLI `pyright` tool 

Pyright CLI tool version: `1.1.357`

**Comment by erictraut** [comment]
Pyright's behavior here is intended, so I don't consider this a bug.

When unpacking an iterable of indeterminate length (as you're doing here), a static type checker needs to make certain assumptions. The assumptions that pyright makes here are consistent with other Python static analysis tools like mypy. The assumption is that both `a` and `b` will receive values from the unpack operator and `_` will receive the `None` value.

It would be possible to change the assumptions, but a change of this type would create other false positive errors and would be disruptive for pyright users who have code bases that are consistent with the current assumptions, so this isn't something I'm interested in changing.

The `reportUnnecessaryComparison` check is a strict-mode check, so I presume you've manually enabled it or have enabled "strict" type checking mode. In strict mode, you may need to make accommodations in your code so as to work with the type checker. I'll leave it to you to figure out how to modify your code accordingly.
